pkg/sql/tablemetadatacache/update_table_metadata_cache_job.go

```diff
@@ -152,18 +171,40 @@ func (j *tableMetadataUpdateJobResumer) markAsCompleted(ctx context.Context) {
 
 // updateTableMetadataCache performs a full update of system.table_metadata by collecting
 // metadata from the system.namespace, system.descriptor tables and table span stats RPC.
-func updateTableMetadataCache(ctx context.Context, ie isql.Executor) error {
+func updateTableMetadataCache(
+	ctx context.Context, ie isql.Executor, resumer *tableMetadataUpdateJobResumer,
+) error {
 	updater := newTableMetadataUpdater(ie)
 	if _, err := updater.pruneCache(ctx); err != nil {
 		log.Errorf(ctx, "failed to prune table metadata cache: %s", err.Error())
 	}
 
+	resumer.updateProgress(ctx, .01)
 	// We'll use the updated ret val in a follow-up to update metrics and
 	// fractional job progress.
-	_, err := updater.updateCache(ctx)
+	_, err := updater.updateCache(ctx, resumer.onBatchUpdate())
 	return err
 }
 
+// onBatchUpdate returns an onBatchUpdateCallback func that updates the progress of the job.
+// The call to updateProgress doesn't happen on every invocation, and only happens every nth
+// invocation, where n is defined by batchesPerProgressUpdate. This is done because each
+// batch update is expected to execute quickly, and updating progress at a high velocity
+// doesn't seem worth it.
+func (j *tableMetadataUpdateJobResumer) onBatchUpdate() onBatchUpdateCallback {
+	batchNum := 0
+	return func(ctx context.Context, totalRowsToUpdate int, rowsUpdated int) {
+		batchNum++
+		estimatedBatches := int(math.Ceil(float64(totalRowsToUpdate) / float64(tableBatchSize)))
+		if batchNum == estimatedBatches {
+			j.updateProgress(ctx, .99)
+		} else if batchNum%batchesPerProgressUpdate == 0 && estimatedBatches > 0 {
+			progress := float32(rowsUpdated) / float32(totalRowsToUpdate)
+			j.updateProgress(ctx, progress)
+		}
+	}
+}
```

Comment:
Can we move the batchNum to the callback and have the updater track that? Then this fn can just be the progress update logic instead of having to return a function.
```
func (j *tableMetadataUpdateJobResumer) onBatchUpdate(
	ctx context.Context, batchNum, totalRowsToUpdate, rowsUpdated int,
) {
	estimatedBatches := int(math.Ceil(float64(totalRowsToUpdate) / float64(tableBatchSize)))
	if batchNum == estimatedBatches {
		j.updateProgress(ctx, .99)
	} else if batchNum%batchesPerProgressUpdate == 0 && estimatedBatches > 0 {
		progress := float32(rowsUpdated) / float32(totalRowsToUpdate)
		j.updateProgress(ctx, progress)
	}
}
```

Reply:
I refactored this PR to no longer use an onBatchUpdate call back and baked this code directly into the `updateCache` function